crates/sui-sdk/src/apis.rs

```diff
@@ -1133,39 +1136,49 @@ impl QuorumDriverApi {
     ) -> SuiRpcResult<SuiTransactionBlockResponse> {
         let (tx_bytes, signatures) = tx.to_tx_bytes_and_signatures();
         let request_type = request_type.unwrap_or_else(|| options.default_execution_request_type());
-        let mut retry_count = 0;
+
         let start = Instant::now();
-        while retry_count < WAIT_FOR_LOCAL_EXECUTION_RETRY_COUNT {
-            let response: SuiTransactionBlockResponse = self
-                .api
-                .http
-                .execute_transaction_block(
-                    tx_bytes.clone(),
-                    signatures.clone(),
-                    Some(options.clone()),
-                    Some(request_type.clone()),
-                )
-                .await?;
+        let response = self
+            .api
+            .http
+            .execute_transaction_block(
+                tx_bytes.clone(),
+                signatures.clone(),
+                Some(options.clone()),
+                Some(request_type.clone()),
+            )
+            .await?;
 
-            match request_type {
-                ExecuteTransactionRequestType::WaitForEffectsCert => {
-                    return Ok(response);
-                }
-                ExecuteTransactionRequestType::WaitForLocalExecution => {
-                    if let Some(true) = response.confirmed_local_execution {
-                        return Ok(response);
-                    } else {
-                        // If fullnode executed the cert in the network but did not confirm local
-                        // execution, it must have timed out and hence we could retry.
-                        retry_count += 1;
-                    }
+        if let ExecuteTransactionRequestType::WaitForEffectsCert = request_type {
+            return Ok(response);
+        }
+
+        // JSON-RPC ignores WaitForLocalExecution, so simulate it by polling for the transaction.
+        let mut poll_response = tokio::time::timeout(WAIT_FOR_LOCAL_EXECUTION_TIMEOUT, async {
+            // Apply a short delay to give state sync a chance to catch up
+            tokio::time::sleep(WAIT_FOR_LOCAL_EXECUTION_DELAY).await;
+
+            let mut interval = tokio::time::interval(WAIT_FOR_LOCAL_EXECUTION_INTERVAL);
+            loop {
+                interval.tick().await;
+
+                if let Ok(poll_response) = self
+                    .api
+                    .http
+                    .get_transaction_block(*tx.digest(), Some(options.clone()))
+                    .await
+                {
+                    break poll_response;
                 }
             }
-        }
-        Err(Error::FailToConfirmTransactionStatus(
-            *tx.digest(),
-            start.elapsed().as_secs(),
-        ))
+        })
+        .await
+        .map_err(|_| {
+            Error::FailToConfirmTransactionStatus(*tx.digest(), start.elapsed().as_secs())
```

Comment:
fwiw, I decided to swallow errors from polling for the transaction to become available in the TS SDK.  In rust, its a lot more common to properly handle different error cases properly, but might be worth thinking though explicitly.  I would suspect that failing to poll for the transaction is significantly more likely to fail (even if its still rare) than local execution failing because of the extra network requests.

In the TS SDK the concern was that application would treat erroring during the polling phase would be handle as the transaction itself failing to execute, and assuming that the transaction was not processed

Reply:
So this code is also swallowing errors during polling. This final error is the one that happens if after 60 seconds, we still weren't able to get a definitive response from the fullnode about the transaction.